client-runtime/src/main/java/com/microsoft/rest/v2/protocol/HttpResponseDecoder.java

```diff
@@ -292,6 +296,50 @@ private Object deserializeHeaders(HttpHeaders headers) throws IOException {
             } else {
                 final String headersJsonString = serializer.serialize(headers, SerializerEncoding.JSON);
                 Object deserializedHeaders = serializer.deserialize(headersJsonString, deserializedHeadersType, SerializerEncoding.JSON);
+
+                final Class<?> deserializedHeadersClass = TypeToken.of(deserializedHeadersType).getRawType();
+                final Field[] declaredFields = deserializedHeadersClass.getDeclaredFields();
+                for (final Field declaredField : declaredFields)
+                {
+                    if (declaredField.isAnnotationPresent(HeaderCollection.class))
+                    {
+                        final Type declaredFieldType = declaredField.getGenericType();
+                        if (TypeToken.of(declaredField.getType()).isSubtypeOf(Map.class))
+                        {
+                            final Type[] mapTypeArguments = getTypeArguments(declaredFieldType);
+                            if (mapTypeArguments.length == 2 && mapTypeArguments[0] == String.class && mapTypeArguments[1] == String.class)
+                            {
+                                final HeaderCollection headerCollectionAnnotation = declaredField.getAnnotation(HeaderCollection.class);
+                                final String headerCollectionPrefix = headerCollectionAnnotation.value().toLowerCase();
+                                final int headerCollectionPrefixLength = headerCollectionPrefix.length();
+                                if (headerCollectionPrefixLength > 0) {
+                                    final Map<String,String> headerCollection = new HashMap<>();
+                                    for (final HttpHeader header : headers)
+                                    {
+                                        final String headerName = header.name();
+                                        if (headerName.toLowerCase().startsWith(headerCollectionPrefix)) {
+                                            headerCollection.put(headerName.substring(headerCollectionPrefixLength), header.value());
+                                        }
+                                    }
+
+                                    final boolean declaredFieldAccessibleBackup = declaredField.isAccessible();
+                                    try {
+                                        if (!declaredFieldAccessibleBackup) {
+                                            declaredField.setAccessible(true);
+                                        }
+                                        declaredField.set(deserializedHeaders, headerCollection);
```

Comment:
Ah, so when we serialize/deserialize we take care of all the fixed-size fields? And then we use reflection to set the HeaderCollection property?

Reply:
Correct.

Reply:
I'll add the HeaderCollection annotation in the generator.